### ISSUE TITLE
Feat: extend type for better module isolation 

### ISSUE THREAD

**Issue by LucienLee** [nestjs/graphql#1433]
## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior

All discussion is under code first schema situation.

In graphql, the resolver chaining is a common pattern to build service. Currently, we can resolve fields by `@ResolveField`, which is good when the sub field is under the same module. However, if the sub field is under another module, the resolver needs to inject the services from other modules in order to resolve the sub field. That's could be fine, but it's not ideal. Because the modules start to couple together.  

For example:  

```ts
// Books/book.model.ts
import { Field, ObjectType, ID } from '@nestjs/graphql';

@ObjectType()
class Book {
  @Field(() => ID)
  id: string;
  
  @Field()
  title: string;
  
  @Field(() => Author)
  author: Author;
}
```

```ts
// Authors/author.model.ts
import { Field, ObjectType, ID } from '@nestjs/graphql';

@ObjectType()
class Author {
  @Field(() => ID)
  id: string;
  
  @Field()
  name: String;
}
```

Now we resolve `Book.author` in the Book resolver as 

```ts
// Books/books.resolver.ts
import { Args, Query, Resolver, ResolveField, Parent } from '@nestjs/graphql';
import { Book } from 'from './models/book.model';
import BooksService from './books.service.ts';
import { Author } from '../Author/models/author.model'; // <- import from author module  
import AuthorsService from '../Author/author.service.ts'; // <- import from author module  

@Resolver(of => Book)
export class BooksResolver {
  constructor(
    private booksService: BooksService,
    private authorsService: AuthorsService, 
  ) {}
  
  @Query(() => Book)
  async book(@Args('id') id: string) {
    return await this.booksService.findOneById({ id });
  }
  
  @ResolveField()
  async author(@Parent() book: Book) {
    const { id } = book;
    return await this.authorsService.findOneById(id);
  }
}

```
The problems are that you do not only need to import other module's models and services in the resolver, you also need to inject them at `book.module.ts`.
You can see that it's hard to decouple these two modules. I don't think that is an elegant way to build a large codebase.

## Expected behavior

In the beginning, I don't want to propose what's the syntax to solve that. I would like to share how did we do with [graphql module](https://graphql-modules.com/) or [graphql tool](https://www.graphql-tools.com/docs/generate-schema#extending-types) and then invite you guys to discuss the solution.

With `makeExecutableSchema` and `extend` syntax from graphql tool. We can do clean module isolation like

```ts
// book module
const typeDefs = gql`
  type Book {
    id: ID!
    title: String!
    author: Author!
  }
`

const bookResolver = {
  Query: {
      book: async (_, { id }) => {
          return await this.booksService.findOneById({ id });
      }
  }
}
```

```ts
// author module
const typeDefs = gql`
  type Author {
    id: ID!
    name: String!
  }
  
  extend type Book {
    author: Author!
  }
`

const authourResolver = {
  Book: {
    author: async (parent) =>  {
        return await this.authorsService.findOneById(parent.id);;
    }
  }
}
```

With extend type, we could decouple two modules cleanly (like federation). Even we remove the author module, the book module could work standalone without any issue. Absolutely, we can't access `book.author` anymore but the other parts are not affected. 

In the previous discussion (such #1146), the contributor proposes we could use abstract type (class) to solve that. The problem is that it's actually still coupling the two modules. By `isAbstract: true` magic, your code would be looks like:

```ts
// Books/book.model.ts
import { Field, ObjectType, ID } from '@nestjs/graphql';

@ObjectType({ isAbstract: true })
class BookBase { // <- how does book module know itself has to be an abstract type with a modified name (here is Base postfix)
  @Field(() => ID)
  id: string;
  
  @Field()
  title: string;
}
```

 ```ts
// Authors/author.model.ts
import { Field, ObjectType, ID } from '@nestjs/graphql';
import { BookBase } from '../Book/models/book.model'; // <- import from book module  

@ObjectType()
class Author {
  @Field(() => ID)
  id: string;
  
  @Field()
  name: String;
}

@ObjectType()
class Book extends BookBase { // <- so if we remove the author module, we don't have book type anymore
   @Field(() => Author)
   author: Author;
}
```
You will found that it's quite strange here. The book model has to know it will be extended, but it should not. Furthermore, the book type declare is assign to the author model. These two modules are still deeply coupled. 

I'm not sure what's the good syntax or solution here. But there is a similar discussion in [type-graphql#228](https://github.com/MichalLytek/type-graphql/issues/228).They propose use `@ObjectType({ extends: true })` like `extend` syntax to solve that. In my opinion, it's great.

## What is the motivation / use case for changing the behavior?

To summarize, I would like to see syntax like `extend` in graphql tool or apollo federation to enable better module isolation, which fits single-responsibility principle and empower nestjs graphql developers to build a larger services

If I misunderstand anything, please point it out. Thanks.

## Environment

<pre><code>
Nest version: 7.6.5
Nest Graphql version: 7.9.8
 
For Tooling issues:
- Node version: v12.13  
- Platform: Mac 
</code></pre>

**Comment by kamilmysliwiec** [comment]
There are no plans to implement this feature in the foreseeable future as we currently don't have enough time to look into that.
